### PR TITLE
Add P2P trigger event and handler

### PR DESCRIPTION
**Pull request by mmpei**
Signed-off-by: peimingming <peimingming@corp.netease.com>

### CODE REVIEW

File: src/controller/event/handler/p2p/preheat.go [hunk 1]
```diff
@@ -0,0 +1,108 @@
...
+	model_tag "github.com/goharbor/harbor/src/pkg/tag/model/tag"
+)
+
+// Handler ...
+type Handler struct {
+	Context func() context.Context
+}
+
+// Handle ...
+func (p *Handler) Handle(value interface{}) error {
```
Comment (steven-zou) [inline src/controller/event/handler/p2p/preheat.go:34]:
```go
switch fmt.Sprintf("%T", value){
  case "*event.PushArtifactEvent":
  case "*event.ScanImageEvent":
  case "*event.ArtifactLabeledEvent":
  default:
  // error
}
```

File: src/controller/event/handler/p2p/preheat.go [hunk 2]
```diff
@@ -0,0 +1,108 @@
...
+	if arti.Type != image.ArtifactTypeImage {
+		return nil
+	}
+	log.Debugf("preheat artifact labeled %s:%s", arti.Artifact.RepositoryName, arti.Artifact.Digest)
+
+	_, err = preheat.Enf.PreheatArtifact(p.Context(), arti)
+	return err
+}
+
+func (p *Handler) convertPushEventToArti(event *event.PushArtifactEvent) *artifact.Artifact {
```
[the remark this answers is not in the document]
Comment (mmpei) [reply src/controller/event/handler/p2p/preheat.go:87]: it is already Artifact and contains all info of artifact.

File: src/controller/event/handler/p2p/preheat.go [hunk 3]
```diff
@@ -0,0 +1,108 @@
...
+			Tag: model_tag.Tag{
+				RepositoryID: event.Artifact.RepositoryID,
+				ArtifactID:   event.Artifact.ID,
+				Name:         t,
+			},
+		}
+		tags = append(tags, tt)
+	}
+
+	return &artifact.Artifact{
```
[the remark this answers is not in the document]
Comment (mmpei) [reply src/controller/event/handler/p2p/preheat.go:100]: No labels when a push event trigger. all info is in Artifact.

File: src/controller/event/handler/p2p/preheat.go [hunk 4]
```diff
@@ -0,0 +1,108 @@
...
+	_, err := preheat.Enf.PreheatArtifact(p.Context(), art)
+	return err
+}
+
+func (p *Handler) handleImageScanned(event *event.ScanImageEvent) error {
+	log.Debugf("preheat image scanned %s:%s", event.Artifact.Repository, event.Artifact.Tag)
+	art, err := p.convertScanEventToArti(event)
+	if err != nil {
+		return err
+	}
```
[the remark this answers is not in the document]
Comment (mmpei) [reply src/controller/event/handler/p2p/preheat.go:71]: at the begining of the function.

File: src/controller/event/handler/p2p/preheat.go (hunk 1 above)
Comment (mmpei) [reply src/controller/event/handler/p2p/preheat.go:34]: ok

File: src/controller/event/handler/p2p/preheat.go (hunk 3 above)
[the remark this answers is not in the document]
Comment (steven-zou) [reply src/controller/event/handler/p2p/preheat.go:101]: if it is an overridden push?

**Pull request closed without merging**